### PR TITLE
Add `CosineLR`: a composable, multiplicative cosine learning rate scheduler

### PR DESCRIPTION
**Pull request by apehex**
## Motivation

The existing `CosineAnnealingLR` scheduler computes **absolute** learning rates using `base_lr` and `eta_min`. This makes it difficult to compose with other schedulers via `ChainedScheduler`, because it overwrites the current learning rate rather than scaling it. Users who want a cosine warmup followed by exponential decay, for example, cannot simply chain the two - the cosine scheduler ignores whatever the exponential scheduler has done to `group["lr"]`.

This PR adds `CosineLR`, a **multiplicative** cosine scheduler that mirrors the design of `LinearLR`: it interpolates a scaling factor between `start_factor` and `end_factor` over `total_iters` steps, then holds steady. Because it applies a relative factor to the current learning rate, it composes correctly with any other scheduler.

This is a concrete step toward the internal consistency goals described in #168044.

## Design

`CosineLR` uses the same parameter names and conventions as `LinearLR`:

| Parameter | Meaning |
|---|---|
| `start_factor` | Multiplicative factor applied at epoch 0 (must be in `(0, 1]`) |
| `end_factor` | Multiplicative factor reached at `total_iters` (must be in `[0, 1]`) |
| `total_iters` | Number of steps over which the factor interpolates |

The factor at step $t$ follows:

$$\gamma_t = \frac{\gamma_s + \gamma_e}{2} + \frac{\gamma_s - \gamma_e}{2} \cos \left(\frac{t \cdot \pi}{T}\right)$$

The learning rate is updated recursively as:

$$\text{lr}\_t = \text{lr}\_{t-1} \cdot \frac{\gamma_t}{\gamma_{t-1}}$$

A closed-form path is provided for the deprecated `step(epoch)` interface:

$$\text{lr}\_t = \text{base\\_lr} \cdot \gamma_t$$

Unlike `CosineAnnealingLR`:
- The schedule is **relative** - it scales `group["lr"]` rather than replacing it
- It supports both **warmup** (`end_factor > start_factor`) and **decay** (`end_factor < start_factor`)
- It composes correctly with `ChainedScheduler`, `SequentialLR`, and compound scheduler lists

## Changes

**`torch/optim/lr_scheduler.py`**
- Added the `CosineLR` class with `get_lr`, `_compute_factor`, and `_get_closed_form_lr`
- Added `CosineLR` to `__all__`

**`test/optim/test_lrscheduler.py`**
- Added `CosineLR` to the imports
- Added dedicated tests:
  - `test_cosinelr` - basic cosine decay against closed-form targets
  - `test_cosinelr_warmup` - increasing schedule (`end_factor > start_factor`)
  - `test_cosinelr_start_factor_limits` / `test_cosinelr_end_factor_limits` - input validation
  - `test_get_last_lr_cosinelr` - `get_last_lr()` correctness
  - `test_cosinelr_with_epoch` - deprecated `step(epoch)` path
  - `test_closed_form_cosinelr` - recursive vs closed-form equivalence over 20 epochs
  - `test_cosinelr_is_constant_for_constant_epoch` - repeated `step(2)` doesn't drift
  - `test_cosinelr_state_dict` - save/load round-trip
  - `test_compound_cosinelr_and_exp_lr` - composition with `ExponentialLR`
  - `test_compound_cosinelr_and_step_lr` - composition with `StepLR`
- Added `CosineLR` to the parametrized test lists for `test_lr_scheduler_state_dict_load`, `test_constant_initial_lr`, and `test_lr_scheduler_checkpoint`

### CODE REVIEW

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -1333,6 +1333,170 @@ def _get_closed_form_lr(self) -> list[float | Tensor]:
...
+        .. math::
+            \gamma_t = \frac{1}{2}\left(\gamma_s + \gamma_e\right)
+            + \frac{1}{2}\left(\gamma_s - \gamma_e\right)
+            \cos\!\left(\frac{t \, \pi}{T}\right)
+        """
+        return (
+            0.5 * (self.start_factor + self.end_factor)
+            + 0.5 * (self.start_factor - self.end_factor)
+            * math.cos(math.pi * step / self.total_iters)
+        )
```
[the remark this answers is not in the document]
Comment (apehex) [reply torch/optim/lr_scheduler.py]: The divisor is now sanitized with `max(1, self.total_iters)`.
I also made sure both `self.start_factor` and `self.end_factor` are strictly positive, hence the factor is always strictly positive for all the steps.

File: test/optim/test_lrscheduler.py [hunk 2]
```diff
@@ -564,6 +677,15 @@ def test_closed_form_poly_lr(self):
...
 
+    def test_closed_form_cosinelr(self):
+        scheduler = CosineLR(
+            self.opt, start_factor=0.3, end_factor=0.7, total_iters=6
+        )
+        closed_form_scheduler = CosineLR(
+            self.opt, start_factor=0.3, end_factor=0.7, total_iters=6
+        )
+        self._test_against_closed_form(scheduler, closed_form_scheduler, 20)
+
```
[the remark this answers is not in the document]
Comment (apehex) [reply test/optim/test_lrscheduler.py:682]: This issue is spanning the whole file: other tests on the closed form use the same scheme.
The process would need to be refactored for all the tests, which is beyond the scope of this PR.

**Comment by apehex** [comment]
I added a few commits to:
- make sure the format agrees with the linter
- and remove the edge case where the relative factor could be null

File: torch/optim/lr_scheduler.py [hunk 3]
```diff
@@ -1333,6 +1333,170 @@ def _get_closed_form_lr(self) -> list[float | Tensor]:
...
         ]
 
+class CosineLR(LRScheduler):
+    r"""Scale the learning rate of each parameter group using a cosine interpolation.
+
+    The multiplicative factor changes from ``start_factor`` to ``end_factor``
+    following a cosine curve over ``total_iters`` steps. Unlike
+    :class:`CosineAnnealingLR`, this scheduler applies a **relative** factor to
+    the current learning rate, making it fully composable with
+    :class:`ChainedScheduler` and other multiplicative schedulers.
```
[the remark this answers is not in the document]
Comment (apehex) [reply torch/optim/lr_scheduler.py:1345]: I added the scheduler to the list `__all__`.

**Comment by janeyx99** [comment]
Hi thanks for the PR. We're in the middle of doing an investigation on our current LRSchedulers to see how we can revamp the system to fix the weird interactions, especially the ones described in the linked RFC. In the process, we don't want to add new LRSchedulers for now, though ofc if there's anything from within PyTorch that's preventing development of this outside the library, please file an issue!

**Comment by apehex** [comment]
Hi @janeyx99 , it makes a lot of sense, no problem!
I'm a factorization addict myself, tell me if you'd like help on this :)
